Minimal set of changes to get FreeBSD stdenv building again

rhelmot opened the pull request:
## Description of changes

- Subset of https://github.com/NixOS/nixpkgs/pull/254801 actually small enough to be reviewable, hopefully

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [X] x86_64-freebsd
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rhelmot committed:
Minimal set of changes to get FreeBSD stdenv building again

Mic92 commented on lib/systems/doubles.nix:
```diff
@@ -13,7 +13,7 @@ let
 
     # FreeBSD
-    "i686-freebsd13" "x86_64-freebsd13"
+    "i686-freebsd" "x86_64-freebsd"
```
Not sure if this was discussed already before but how is ABI compatibility now preserved? I.e. how will be kernel and userland kept in sync?

rhelmot replied:
This was discussed here: https://github.com/NixOS/nixpkgs/pull/254801#issuecomment-1984834380

Mic92 replied:
Ok. Looks like the FreeBSD community is doing the same with their own tooling:

```
poudriere is a BSD-licensed utility for creating and testing FreeBSD packages. It uses FreeBSD jails to set up isolated compilation environments. These jails can be used to build packages for versions of FreeBSD that are different from the system on which it is installed, and also to build packages for i386 if the host is an amd64 system. Once the packages are built, they are in a layout identical to the official mirrors. These packages are usable by [pkg(8)](https://man.freebsd.org/cgi/man.cgi?query=pkg&sektion=8&format=html) and other package management tools.
```

Mic92 commented on maintainers/maintainer-list.nix:
```diff
@@ -1647,6 +1647,7 @@
     email = "me@artem.ist";
     github = "artemist";
     githubId = 1226638;
+    matrix = "@artemist:mildlyfunctional.gay";
```
Looks like this was added by accident?

```suggestion
```

rhelmot replied:
not quite... it was added by artemis, who has been collaborating with me on this project, on our private staging branch, but I suppose this diff does remove the attribution for that line, which is probably important for the maintainers file.

Mic92 commented on pkgs/build-support/setup-hooks/make-binary-wrapper/default.nix:
```diff
@@ -10,7 +10,7 @@
 makeSetupHook {
   name = "make-binary-wrapper-hook";
-  propagatedBuildInputs = [ dieHook ];
+  propagatedBuildInputs = [ dieHook cc ];
```
Is this to make it work in stdenvNoCC?

rhelmot replied:
@artemst can you comment? here's the original commit https://github.com/rhelmot/nixpkgs/commit/4e0769e941b81817c0372cf833d73476ea92e756

rhelmot replied:
@artemist can you comment? here's the original commit https://github.com/rhelmot/nixpkgs/commit/4e0769e941b81817c0372cf833d73476ea92e756

artemist replied:
Yes, the tests and several normal packages use makeBinaryWrapper in runCommand. This was a simpler change than making all of them runCommandCC

rhelmot committed:
Update maintainers/maintainer-list.nix

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

Mic92 commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -71,6 +74,8 @@ python3.pkgs.buildPythonApplication rec {
     libxcrypt
   ];
 
+  propagatedNativeBuildInputs = lib.optionals stdenv.isFreeBSD [ buildPackages.freebsd.ldd ];
```
Are we bound in freebsd to a single linker? For other platforms this is part of stdenv and can be overwritten.

rhelmot replied:
ldd is the dynamic linker, it's used here to print the runtime dependencies of stuff, are you thinking of ld?

Mic92 replied:
Ah. Sorry. Mixed them up. Yeah `ldd` is always libc specific by design.

Mic92 commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -88,8 +93,16 @@ python3.pkgs.buildPythonApplication rec {
 
+  postPatch = lib.optionalString stdenv.hostPlatform.isFreeBSD ''
+    # This edge case is explicilty part of meson but is wrong for nix
+    sed -E -i -e s/mesonlib.is_freebsd../False/g mesonbuild/modules/pkgconfig.py
```
Fine to patch it this way, if it's a failure also visible at build time.

rhelmot replied:
The failure manifests at build time not for meson but for packages using meson. Is that what you meant?

Mic92 replied:
Ok. Ideally we would be not able to build the package at all if the code changes but the error cannot be captured.

Mind using this patch instead? Afaik this can be even used unconditionally.

```
diff --git a/mesonbuild/modules/pkgconfig.py b/mesonbuild/modules/pkgconfig.py
index ebe0d92d5..99f84f8e3 100644
--- a/mesonbuild/modules/pkgconfig.py
+++ b/mesonbuild/modules/pkgconfig.py
@@ -693,7 +693,7 @@ def parse_variable_list(vardict: T.Dict[str, str]) -> T.List[T.Tuple[str, str]]:
         pcfile = filebase + '.pc'
         pkgroot = pkgroot_name = kwargs['install_dir'] or default_install_dir
         if pkgroot is None:
-            if mesonlib.is_freebsd():
+            if False:
                 pkgroot = os.path.join(_as_str(state.environment.coredata.get_option(mesonlib.OptionKey('prefix'))), 'libdata', 'pkgconfig')
                 pkgroot_name = os.path.join('{prefix}', 'libdata', 'pkgconfig')
             elif mesonlib.is_haiku():

```

rhelmot replied:
Ohhhhh, I understand what you're saying now.

That's fine, sure. I agree with the aim of using a diff file, and I only made it conditional to avoid mass rebuilds on the linux side.

Mic92 replied:
It also needs to be conditional, is what I am realizing now.

Mic92 commented on pkgs/development/compilers/llvm/16/compiler-rt/freebsd-i386.patch:
```diff
@@ -0,0 +1,21 @@
+ #include <stdbool.h>
+ #include <stdint.h>
+ 
+-// x86_64 FreeBSD prior v9.3 define fixed-width types incorrectly in
```
So now it's defined correctly?

rhelmot replied:
Indeed, and we don't support FreeBSD<14, maybe in the future to be <13. I suspect the problematic part of this patch was just including sys/param.h, but regardless it prevented literally any i386 freebsd code (in this case, the bootloader) from compiling.

Mic92 commented on pkgs/os-specific/bsd/freebsd/by-name/bmakeMinimal.nix:
```diff
@@ -0,0 +1,93 @@
+    # make needs this to pick up our sys make files
+    export NIX_CFLAGS_COMPILE+=" -D_PATH_DEFSYSPATH=\"$out/share/mk\""
+
+    # Copied from netbsd. we need to prevent it from trying to adjust the permissions of installed files.
+    #set +e
+    #find $BSDSRCDIR -print0 -name '*.mk' | while IFS= read -r -d "" f; do
+    #  echo $f
+    #  substituteInPlace "$f" \
+    #    --replace '-o ''${DOCOWN}' "" \
+    #    --replace '-g ''${DOCGRP}' "" \
+    #    --replace '-o ''${BINOWN}' "" \
+    #    --replace '-g ''${BINGRP}' "" \
+    #    --replace '-o ''${RUMPBINOWN}' "" \
+    #    --replace '-g ''${RUMPBINGRP}' "" \
+    #    --replace '-o ''${DEBUGOWN}' "" \
+    #    --replace '-g ''${DEBUGGRP}' "" \
+    #    --replace '-o ''${KMODULEOWN}' "" \
+    #    --replace '-g ''${KMODULEGRP}' "" \
+    #    --replace '-o ''${LIBOWN}' "" \
+    #    --replace '-g ''${LIBGRP}' "" \
+    #    --replace '-o ''${MANOWN}' "" \
+    #    --replace '-g ''${MANGRP}' "" \
+    #    --replace '-o ''${NLSOWN}' "" \
+    #    --replace '-g ''${NLSGRP}' ""
+    #done
+    #set -e
```
Is this dead code?

rhelmot replied:
It is dead! I kept it around because I wasn't sure if it was right and then forgot about it until it was mass rebuild time :( I will eat that rebuild now

Mic92 commented on pkgs/os-specific/bsd/freebsd/by-name/bmakeMinimal.nix:
```diff
@@ -0,0 +1,93 @@
+  buildPhase = ''
+    runHook preBuild
+
+    sh ./make-bootstrap.sh
```
Is assume in these type of derivation from bsd or is it bash?

rhelmot replied:
sh is assumed to be bash in all the bsd derivations

Mic92 commented on pkgs/os-specific/bsd/freebsd/by-name/filterSource.nix:
```diff
@@ -0,0 +1,11 @@
+{ lib, pkgsBuildBuild, runCommand, writeText, source }:
+{ pname, path, extraPaths ? [] }:
+let
+  sortedPaths = lib.naturalSort ([ path ] ++ extraPaths);
```
Does rsync wants these paths to be sorted? Or is this to avoid rebuilds?

rhelmot replied:
Genuinely unsure. Another @artemist commit: https://github.com/rhelmot/nixpkgs/commit/b4b9e450a3b74d4437ddbb5e5fe2ec5b0ef24805

Maybe it was to make it possible to binary search some issue which was coming up during filtering?

artemist replied:
It's not technically required, but the rsync man page says "NOTE: sorting the list of files in the --files-from input helps rsync to be more efficient, as it will avoid re-visiting the path elements that are shared between adjacent entries."

This isn't generally a problem but we do sometimes have both a child and its parent in the list. It also has the added benefit of reducing rebuilds as you've mentioned.

rhelmot committed:
Remove dead code

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+    ref_type: str,
+    supported_refs: typing.List[str],
+    old_versions: dict,
+):
```
```suggestion
) -> dict[str, Any]:
```

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+    ref_name: str,
+    ref_type: str,
+    supported_refs: typing.List[str],
+    old_versions: dict,
```
```suggestion
    old_versions: dict[str, str],
```

Mic92 replied:
Those type hints are just guesses on my side.

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+    rev: git.objects.commit.Commit,
+    ref_name: str,
+    ref_type: str,
+    supported_refs: typing.List[str],
```
```suggestion
    supported_refs: list[str],
```

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+    return list(df["Branch"])
+
+
+def query_version(repo: git.Repo):
```
```suggestion
def query_version(repo: git.Repo) -> dict[str, Any]:
```

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+)
+
+
+def request_supported_refs() -> typing.List[str]:
```
```suggestion
def request_supported_refs() -> list[str]:
```

Mic92 commented on pkgs/os-specific/bsd/freebsd/update.py:
```diff
@@ -0,0 +1,178 @@
+    }
+
+
+# Normally uses /run/user/*, which is on a tmpfs and too small
```
I leave this up to you, but putting all code in a `main()` function helps to avoid accidentally introducing module-wide global variables.

Mic92 commented on pkgs/stdenv/freebsd/always-patchelf.sh:
```diff
@@ -0,0 +1,5 @@
+fixupOutputHooks+=(_FreeBSDPatchelfShrink)
+
+_FreeBSDPatchelfShrink() {
+    find $prefix -type f | xargs -n1 patchelf --shrink-rpath &>/dev/null || true
```
Maybe only executable files need to be patchelf'd?

rhelmot replied:
not necessarily - shared objects are often -x and need to be patchelfed.

rhelmot committed:
FreeBSD update.py: Add type annotation

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

rhelmot committed:
Update pkgs/os-specific/bsd/freebsd/update.py: add type annotation

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

Mic92 commented on pkgs/stdenv/freebsd/autotools-abspath.sh:
```diff
@@ -0,0 +1,11 @@
+
+autotoolsAbspathHook() {
+    if [ -z "$dontAutopatchAutotools" ]; then
+        AUTOTOOLS_FILES="configure config.guess"
+        for filename in $AUTOTOOLS_FILES; do
+            find . -name $filename -exec \
+                sed -e 's@/usr/bin/uname@uname@g' -i '{}' ';'
+        done
```
I guess this list doesn't get much longer? And we should only sed files.

```suggestion
        find . \( -name 'configure' -o -name 'config.guess' \) -type f -exec \
            sed -e 's@/usr/bin/uname@uname@g' -i '{}' ';'
```

rhelmot committed:
FreeBSD update.py: Add type annotations

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

Mic92 commented on pkgs/stdenv/freebsd/bootstrap-files/bootstrapFiles.nix:
```diff
@@ -0,0 +1,99 @@
+{ system }: ((import <nixpkgs> {}).callPackage ({ stdenv, pkgsCross, runCommand, lib, stdenvAdapters, buildPackages, ... }:
```
Supposedly we need to make sure that hydra provides these bootstrap files soon?
Needs to be coordinated with the hydra team.

Mic92 commented on pkgs/stdenv/freebsd/default.nix:
```diff
@@ -1,276 +1,460 @@
+    # These bootstrap hashes generated by compiling from nixpkgs 46c6e8a75f2329633858ca13a589698fcbf6231c on x86_64-linux
+    x86_64-freebsd = {
+      bash = {
+        url = "https://f005.backblazeb2.com/file/nixbsd/aldkfyqiq0nspljgl5pr0pgvxq99qirk-x86_64-freebsd-bash";
```
We can even replace this with NixOS foundation hosted binaries before this PR is merged, I think. I will ask.

Mic92 replied:
Ok. So we need someone that can upload to tarballs.nixos.org i.e. @lovesegfault
Furthermore in the ideal case, we could bootstrap this tarball from an existing nixpkgs target for example Linux. Is this possible?

rhelmot replied:
as indicated by the comment, the tarball was bootstrapped from an existing nixpkgs target, x86_64-linux :)

The scripts for building the bootstrap tarball and generating this config are committed right now. I don't know if there's a kosher way to integrate them with hydra, but I'll gladly make it work.

Mic92 replied:
If we can cross-compile them, than we can manually upload the first version and for future versions we have scripts to regenerate bootstrap tarballs in nixpkgs.

Mic92 replied:
I contacted @lovesegfault again on matrix in the hopes, he would respond.

rhelmot committed:
lint update.py

Mic92 commented on pkgs/stdenv/freebsd/default.nix:
```diff
@@ -1,276 +1,460 @@
+      inherit (all-bootstrap-files) tar unxz mkdir chmod bootstrapFiles;
+    });
+
+    linkBS = (attrs: derivation (attrs // {
```
Interesting name choice here ;)

rhelmot replied:
it's how i feel about the pure bootstrapping process -_-

rhelmot committed:
Update pkgs/stdenv/freebsd/autotools-abspath.sh: make enumeration cleaner

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

Mic92 reviewed (approved):
Generally I am ok with the patches above. Given the large size things might have slipped through. I would appreciate a second pair of eyes.

rhelmot committed:
meson: convert pkgconfig patch to diff

rhelmot committed:
chore: replace freebsd13 with freebsd globally

rhelmot committed:
freebsd: include some patches which didn't make it into the merge

SuperSandro2000 commented:
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

Also please fix eval-lib-tests. I think it is failing at https://github.com/NixOS/nixpkgs/blob/ebaf43d1633e9783c4681e9632841153dc0335b0/lib/tests/systems.nix#L55

SuperSandro2000 commented on pkgs/by-name/me/meson/package.nix:
```diff
@@ -68,6 +68,10 @@ python3.pkgs.buildPythonApplication rec {
     # Nixpkgs cctools does not have bitcode support.
     ./006-disable-bitcode.patch
+
+  ] ++ lib.optionals stdenv.isFreeBSD [
```
Can we make this unconditional somehow? otherwise it could easily stop applying on update.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20744,7 +20747,7 @@ with pkgs;
 
   cyrus_sasl = callPackage ../development/libraries/cyrus-sasl {
-    libkrb5 = if stdenv.isFreeBSD then heimdal else libkrb5;
+    #libkrb5 = if stdenv.isFreeBSD then heimdal else libkrb5;
```
If this line is no longer desired, we should just delete it.

SuperSandro2000 commented on pkgs/tools/security/rhash/dont-fail-ln.patch:
```diff
@@ -0,0 +1,8 @@
++++ b/librhash/Makefile
+@@ -32,1 +32,1 @@
+-	  ln -s $(LIBRHASH_SHARED) $(LIBDIR)/$(LIBRHASH_SO_MAJ)
++	  ln -s $(LIBRHASH_SHARED) $(LIBDIR)/$(LIBRHASH_SO_MAJ) || true
```
I am not sure if that is so nice on all other platforms.

SuperSandro2000 commented on pkgs/tools/misc/expect/default.nix:
```diff
@@ -29,12 +30,19 @@ tcl.mkTclDerivation rec {
     sed -i "s,/bin/stty,$(type -p stty),g" configure.in
   '';
 
+  # software doesn't work on clang16 yet...
+  preConfigure = ''
+    export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -Wno-implicit-int -Wno-implicit-function-declaration"
+  '';
```
```suggestion
  # software doesn't support clang16 yet...
  env.NIX_CFLAGS_COMPILE = "-Wno-implicit-function-declaration -Wno-implicit-int";
```

SuperSandro2000 commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -16,7 +16,8 @@
 , withPrefix ? false
-, singleBinary ? "symlinks" # you can also pass "shebangs" or false
+, singleBinary ? if stdenv.isFreeBSD then false else "symlinks" # you can also pass "shebangs" or false
+, withAutoreconf ? true,
```
We shouldn't add such an option.

Mic92 replied:
Don't we need to disable this on stdenv bootstrapping? I don't think our bootstrapping tarball has enough dependencies to have this.

rhelmot replied:
This is an incredibly old diff, from back when I was doing the impure bootstrap. it's distinctly possible that we can use autoreconf now. I will try.

rhelmot replied:
It is removed now.

SuperSandro2000 commented on pkgs/tools/misc/coreutils/default.nix:
```diff
@@ -84,9 +85,9 @@ stdenv.mkDerivation rec {
-  nativeBuildInputs = [
-    # autoreconfHook is due to patch, normally only needed for cygwin
-    autoreconfHook
+  # autoreconfHook is due to patch, normally only needed for cygwin
+  nativeBuildInputs = optionals withAutoreconf [ autoreconfHook ]
+  ++ [
```
It would be easier to just keep this on all platforms

Mic92 replied:
See above.

rhelmot committed:
meson: make freebsd patch unconditional

rhelmot committed:
expect: cleanup - move NIX_CFLAGS_COMPILE into env package env declaration

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

rhelmot committed:
cyrus_sasl: cleanup - remove commented line

SuperSandro2000 commented on pkgs/stdenv/freebsd/default.nix:
```diff
@@ -1,276 +1,460 @@
+      buildPlatform = localSystem;
+      hostPlatform = localSystem;
+      targetPlatform = localSystem;
+      initialPath = [ prevStage.coreutils prevStage.gnutar prevStage.findutils prevStage.gnumake prevStage.gnused prevStage.patchelf prevStage.gnugrep prevStage.gawk prevStage.diffutils prevStage.patch prevStage.bash prevStage.gzip prevStage.bzip2 prevStage.xz bsdcp];
```
We might want to break this into multiple lines.

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/by-name/mkDerivation.nix:
```diff
@@ -0,0 +1,60 @@
+  } // attrs.meta or {};
+} // lib.optionalAttrs stdenv'.hasCC {
+  # TODO should CC wrapper set this?
+  CPP = "${stdenv'.cc.targetPrefix}cpp";
+} // lib.optionalAttrs stdenv'.isDarwin {
+  MKRELRO = "no";
+} // lib.optionalAttrs (stdenv'.cc.isClang or false) {
+  HAVE_LLVM = lib.versions.major (lib.getVersion stdenv'.cc.cc);
+} // lib.optionalAttrs (stdenv'.cc.isGNU or false) {
+  HAVE_GCC = lib.versions.major (lib.getVersion stdenv'.cc.cc);
+} // lib.optionalAttrs (stdenv'.isx86_32) {
+  USE_SSP = "no";
```
We could simplify those conditions by using env.

rhelmot replied:
This code actually predates me - see https://github.com/NixOS/nixpkgs/blob/a7d38622e750d08f14de9c671caf88976b54ad00/pkgs/os-specific/bsd/freebsd/default.nix#L139-L153

Do you still want it changed?

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/by-name/localedef.nix:
```diff
@@ -0,0 +1,33 @@
+{ patchesRoot, mkDerivation, buildPackages, buildFreebsd, lib, stdenv, compat, ... }:
```
We shouldn't be required to use ... in all those packages.

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/by-name/libncurses-tinfo.nix:
```diff
@@ -0,0 +1,16 @@
+    make $makeFlags "CFLAGS=-D_VA_LIST -D_VA_LIST_DECLARED -Dva_list=__builtin_va_list -I$BSDSRCDIR/contrib/ncurses/ncurses -I$BSDSRCDIR/contrib/ncurses/include -I." ncurses_dll.h make_hash make_keys
+  '';
+
+  #makeFlags = [
```
Can this be removed?

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/by-name/libc.nix:
```diff
@@ -0,0 +1,263 @@
+    make -C $BSDSRCDIR/lib/libc_nonshared $makeFlags install
+
+    mkdir $BSDSRCDIR/lib/libmd/sys
+    make -C $BSDSRCDIR/lib/libmd $makeFlags
+    make -C $BSDSRCDIR/lib/libmd $makeFlags install
```
Can't this be deduplicated a lot by using ``for a in b c``?

rhelmot replied:
It can be, but I find it a lot easier to enumerate the options fully, especially when there are neighboring lines which ought to fit into the loop but have overrides applied.

SuperSandro2000 commented on pkgs/os-specific/bsd/freebsd/by-name/bin.nix:
```diff
@@ -0,0 +1,39 @@
+mkDerivation {
+  pname = "bins";
+  path = "bin";
+  extraPaths = [/*"sbin" "usr.bin" "usr.sbin"*/ "sys/conf" "sys/sys/param.h" "contrib/sendmail" "contrib/tcsh" "usr.bin/printf" "lib/libsm"];
```
Is this more dead code?

rhelmot committed:
coreutils: remove spurious option

SuperSandro2000 commented on pkgs/development/compilers/llvm/16/compiler-rt/default.nix:
```diff
@@ -113,10 +119,17 @@ stdenv.mkDerivation {
       --replace "#include <assert.h>" ""
+  '' + lib.optionalString (useLLVM && !haveLibc && stdenv.hostPlatform.isFreeBSD) ''
+    # As per above, but in FreeBSD assert is a macro and simply allowing it to be implicitly declared causes Issues!!!!!
+    substituteInPlace lib/builtins/clear_cache.c \
+      --replace "#include <assert.h>" "#define assert(e) ((e)?(void)0:__assert(__FUNCTION__,__FILE__,__LINE__,#e))"
+    substituteInPlace lib/builtins/cpu_model.c \
+      --replace "#include <assert.h>" "#define assert(e) ((e)?(void)0:__assert(__FUNCTION__,__FILE__,__LINE__,#e))"
```
```suggestion
    substituteInPlace lib/builtins/clear_cache.c lib/builtins/cpu_model.c \
      --replace "#include <assert.h>" "#define assert(e) ((e)?(void)0:__assert(__FUNCTION__,__FILE__,__LINE__,#e))"
```

rhelmot committed:
stdenv-freebsd: clean up formatting for initialPath

rhelmot committed:
freebsd/by-name: cleanup - remove ... from package arguments

rhelmot committed:
freebsd.bin: remove dead code

rhelmot committed:
llvmPackages16.compiler-rt: reformat substituteInPlace calls

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

rhelmot committed:
freebsd.libncurses-tinfo: remove dead code

rhelmot committed:
freebsd.iconv: init

rhelmot committed:
meson: make freebsd diff patch actually apply

rhelmot committed:
freebsd.libdl: init

rhelmot committed:
FreeBSD bootstrap: remove libcxxabi

rhelmot committed:
libxcrypt: fix build on freebsd/llvm17

rhelmot committed:
llvmPackages17.libcxx: fix FreeBSD vis a vis static compilation

rhelmot committed:
zlib: Fix linker error on freebsd/llvm17

rhelmot committed:
binutils: fix linker error on freebsd/llvm17

rhelmot committed:
ncurses: add a flag to prevent linker errors for FreeBSD cross

rhelmot committed:
stdenv/freebsd: update bootstrap hashes

rhelmot committed:
rhash: make FreeBSD patches less ad-hoc

rhelmot commented:
I realized I forgot to push the changes necessary to make the bootstrap files build cross. I've done that now, which was actually difficult due to the fact that I haven't rebuilt the bootstrap tarball since prior to the switch to llvm17.

I believe all expressed concerns have been either addressed or responded to.

rhelmot committed:
bintools-wrapper: only install existing outputs

rhelmot committed:
freebsd: fix infinite recursion; remove premature drm stuff; add libncurses

rhelmot committed:
bintools-wrapper, cc-wrapper: do not propagate outputsToInstall

* I do not understand why these errors are manifesting in the way they
  are but checkMeta is claiming they are being set to bad values.

rhelmot committed:
Merge remote-tracking branch 'upstream/staging' into freebsd-minimal

Ericson2314 commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -59,6 +59,7 @@
 , postLinkSignHook ? null, signingUtils ? null
 }:
 
+#assert propagateDoc -> bintools ? man;
```
intentional?

rhelmot replied:
I meant to uncomment that! I ran into errors during development that were drastically simplified by this assertion. I've pushed a fix.

Ericson2314 commented:
Maybe we can first review all the things which should work with the existing FreeBSD cross compilation (e.g. new packages, packages fixes), and then review the native building stuff?

alyssais commented:
Even in parallel I think would work, it's just hard to review so many changes in a single sitting.

Ericson2314 commented:
Yeah parallel is best, though I am also sympathetic it is easy to (a) miss one (b) by the time all the pieces are reviewed in isolation, the put-back-together thing doesn't work for some reason again.

rhelmot commented:
It'll be a pain, but I can parallelize this pull request and then fix the aftermath after everything is merged. My goal was definitely to make it so that I had the absolute minimum product I could guarantee would still work (i.e. still hashed to the same value), and with the amount of fudging that was necessary to make it so that CI passed and all the review concerns were satisfied, I don't think things would merge cleanly.

Would it be reasonable instead for me or someone else to subdivide this pull request informally and provide a "review checklist" in the initial post (editable by maintainers) that individual reviewers should check pieces off of as they reviewed them, then we merge it once it's been fully reviewed? That way I can keep ahold of the conflicts and the correctness.

rhelmot committed:
bintools-wrapper: uncomment assertion

alyssais commented:
> Would it be reasonable instead for me or someone else to subdivide this pull request informally and provide a "review checklist" in the initial post (editable by maintainers) that individual reviewers should check pieces off of as they reviewed them

The problem with this approach is that a change is never "done" until it's merged.  People can keep making suggestions throughout the huge PR as long as one thing is not checked off.  It's also extremely difficult to keep track of lots of different conversations on a big PR.

If you need to have an overview of the whole thing, big PR open as a draft showing everything that needed to happen, and so you could verify that put together everything still worked, but actually merge stuff gradually through smaller, reviewable PRs, and then rebase the big PR on top of those every time something is merged.

rhelmot commented:
Okay. What's the granularity of the pull requests you want? I am hoping for "one pull request for every package except for the FreeBSD scope, which can be one chunk". There will be multiple phases and it will be challenging for me to stay on task if it takes longer than a few days, but everything after stdenv should be much more straightforward.

alyssais commented:
The smaller they are, the easier they will be to review, and the more likely it will be that they can be merged quickly.  The main problem I see with the FreeBSD scope is that the changes are not explained at all.  Splitting it into more commits would make it easier to see which parts need explanation, but it could stay one big commit as well.

To manage expectations, though, I don't think we're going to get through this in a few days.  It's just too much code for that to be feasible if any amount of back and forth is required.

rhelmot committed:
Merge branch 'staging' into freebsd-minimal

Mic92 commented:
It's not too bad actually to review. I spent maybe 1/2-1h on it. Also most of the changes are only affecting a new experimental freebsd target anyway, so we don't need to worry too much if it's not perfect the first time.

Ericson2314 commented:
OK now that #298849 is merged, is it time to tackle this one again? :)

rhelmot committed:
Merge remote-tracking branch 'upstream/master' into freebsd-minimal

rhelmot committed:
Merge remote-tracking branch 'upstream/staging' into freebsd-minimal

rhelmot commented:
Since you merged one of my PRs into master and one into staging, I had to merge both of them in order to get this diff to look right. Things should settle down next time staging and master are merged. If you want to review the diff before then you should diff against https://github.com/rhelmot/nixpkgs/commit/ff8f7c5c970a18d3344c8bb678d6eea538c2b675.

I will be traveling until at least Tuesday, so in the interest of accelerating the merge process in hopes of making the 24.05 cutoff, I'm authorizing you (@Ericson2314) to mutate this to your hearts content - this should probably entail opening smaller pull requests for each related set of packages.

I haven't actually been able to test the native freebsd stdenv build yet, on account of it's late and I would need to rebuild the bootstrap files. I have my FreeBSD vm with the jailed nix patches, so I'm in a position to be able to test this pending my return from travel, but you should probably get that same setup if you're interested in testing before then.

rhelmot committed:
freebsd: fix additional merge conflicts

bash: fix freebsd build

rhelmot committed:
stdenv/freebsd: update bootstrap hashes

artemist committed:
stdenv/freebsd: Pass runtimeShell

rhelmot committed:
stdenv/freebsd: finish making things eval

rhelmot committed:
freebsd: do not use compiler with libs for compiling libc

rhelmot committed:
Update bootstrap hashes

rhelmot committed:
stdenv/freebsd: fix eval again

rhelmot committed:
lib: add x86_64-freebsd back to flake-systems

rhelmot committed:
rsync: Make enableLZ4 = false actually work

rhelmot committed:
stdenv/freebsd: provide locales to all stdenvs, not just the final one

rhelmot committed:
rsync: Make enableOpenSSL = false actually work

rhelmot commented:
NOW it works again. I'll work on breaking this into smaller PRs, can someone else work on getting the generate-bootstrap-files.sh to run in a trusted environment so we can upload the results somewhere trusted?

alyssais commented:
> can someone else work on getting the generate-bootstrap-files.sh to run in a trusted environment so we can upload the results somewhere trusted?

The normal way we'd do this would be to add it to make-bootstrap-tools-cross.nix, which is built by Hydra (and which is currently under stdenv/linux, lol, so it looks like it might need some adjustment for FreeBSD)

rhelmot commented:
I've been compiling the bootstrap tools cross, so that should work fine.

rhelmot committed:
llvmPackages: actually use previously-added patches

The patches were only applied on versions they were actually tested on,
but likely apply and fix the same issues on more versions.

rhelmot committed:
freebsd.boot-install: filter the arguments that are not applicable to coreutils install

rhelmot committed:
cmake: remove propagatedBuildInput ps and patch code to use an abspath, as is custom

rhelmot committed:
expect: clean up the patches a bit

rhelmot committed:
coreutils: switch freebsd singleBinary default to shebangs

rhelmot committed:
texinfo: remove unneeded CFLAGS entry

rhelmot committed:
cpython: remove deterministic uname and commit harder to patching uname uses

rhelmot committed:
freebsd: strip installed binaries, remove associated hack from coreutils

rhelmot committed:
re-disable FreeBSD single binary

rhelmot committed:
cpython: make uname patch unconditional

rhelmot committed:
bash: remove a configure flag now that we have a stripped libc

rhelmot committed:
coreutils: make autoreconfHook patch conditional

rhelmot committed:
all-packages: remove some extraneous code to reduce the diff

rhelmot committed:
freebsd.locale: init

rhelmot committed:
Undo the libuuid diff, it's not needed for right now.

rhelmot committed:
meson: clean up dramatically

rhelmot committed:
re2c: undo disable tests

rhelmot committed:
bash: make FreeBSD configure flag less conditional

Mic92 commented on pkgs/applications/system/asusctl/default.nix:
```diff
@@ -3,31 +3,34 @@
 }:
 
 rustPlatform.buildRustPackage rec {
   pname = "asusctl";
```
Is this package diff on purpose in this freebsd branch?

rhelmot replied:
It is not :) ty for noticing.

rhelmot commented:
Note: do not spend review cycles on this one. Everything from it was broken into one of the linked pull requests, go review those!

rhelmot committed:
bintools: drop the jank hunk

rhelmot committed:
make autotools-abspath obsolete

rhelmot committed:
coreutils: enable single-binary build on FreeBSD

artemist committed:
freebsd.libncurses: Fix typo
